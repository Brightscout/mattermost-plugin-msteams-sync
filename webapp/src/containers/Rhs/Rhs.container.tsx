import React, {useCallback, useEffect, useMemo, useState} from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import {useDispatch} from 'react-redux';

import {Button, Input, Spinner} from '@brightscout/mattermost-ui-library';

import {Dialog, Icon, IconName, LinkChannelModal, LinkedChannelCard, Snackbar, WarningCard} from 'components';
import {pluginApiServiceConfigs} from 'constants/apiService.constant';
import {DialogsIds, debounceFunctionTimeLimitInMilliseconds, defaultPage, defaultPerPage} from 'constants/common.constants';
import Constants from 'constants/connectAccount.constants';
import {channelListTitle, noMoreChannelsText} from 'constants/linkedChannels.constants';
import useApiRequestCompletionState from 'hooks/useApiRequestCompletionState';
import useAlert from 'hooks/useAlert';
import useDialog from 'hooks/useDialog';
import usePluginApi from 'hooks/usePluginApi';
import usePreviousState from 'hooks/usePreviousState';
import {getConnectedState, getIsRhsLoading, getLinkModalState, getRefetchState, getSnackbarState} from 'selectors';
import {setConnected} from 'reducers/connectedState';
import utils from 'utils';

import './Rhs.styles.scss';
import {showLinkModal} from 'reducers/linkModal';
import {resetRefetch} from 'reducers/refetchState';

export const Rhs = () => {
    const {makeApiRequestWithCompletionStatus, getApiState, state} = usePluginApi();
    const {Avatar} = window.Components;

    // state variables
    const [totalLinkedChannels, setTotalLinkedChannels] = useState<ChannelLinkData[]>([]);
    const [paginationQueryParams, setPaginationQueryParams] = useState<PaginationQueryParams>({
        page: defaultPage,
        per_page: defaultPerPage,
    });
    const [getLinkedChannelsParams, setGetLinkedChannelsParams] = useState<SearchParams | null>({...paginationQueryParams});
    const {connected, msteamsUserId, username, isAlreadyConnected} = getConnectedState(state);
    const [searchLinkedChannelsText, setSearchLinkedChannelsText] = useState('');
    const [firstRender, setFirstRender] = useState(true);

    const previousState = usePreviousState({searchLinkedChannelsText});

    const dispatch = useDispatch();
    const showAlert = useAlert();

    // Increase the page number by 1
    const handlePagination = () => {
        setPaginationQueryParams({...paginationQueryParams, page: paginationQueryParams.page + 1,
        });
    };

    // Make api call to connect user account
    const connectAccount = useCallback(() => {
        makeApiRequestWithCompletionStatus(pluginApiServiceConfigs.connect.apiServiceName);
    }, []);

    // Make api call to disconnect user account
    const disconnectUser = useCallback(() => {
        makeApiRequestWithCompletionStatus(pluginApiServiceConfigs.disconnectUser.apiServiceName);
    }, []);

    // Reset the pagination params and empty the subscription list
    const resetStates = useCallback(() => {
        setPaginationQueryParams({page: defaultPage, per_page: defaultPerPage});
        setTotalLinkedChannels([]);
    }, []);

    // Check if more linked channels are present after the current page
    const hasMoreLinkedChannels = useMemo<boolean>(() => (
        (totalLinkedChannels.length - (paginationQueryParams.page * defaultPerPage) === defaultPerPage)
    ), [totalLinkedChannels]);

    // Show disconnect dialog component
    const {DialogComponent, showDialog, hideDialog} = useDialog(DialogsIds.disconnect);

    const {isRhsLoading} = getIsRhsLoading(state);
    const {isOpen} = getSnackbarState(state);
    const {refetch} = getRefetchState(state);
    const {data} = getApiState(pluginApiServiceConfigs.whitelistUser.apiServiceName);

    const {presentInWhitelist} = data as WhitelistUserResponse;
    const {data: linkedChannels, isLoading} = getApiState(pluginApiServiceConfigs.getLinkedChannels.apiServiceName, getLinkedChannelsParams as SearchParams);

    // Handle searching of linked channels with debounce
    useEffect(() => {
        if (firstRender) {
            return;
        }

        const timer = setTimeout(() => {
            resetStates();
        }, debounceFunctionTimeLimitInMilliseconds);

        /* eslint-disable consistent-return */
        return () => {
            clearTimeout(timer);
        };
    }, [searchLinkedChannelsText]);

    // Make api call to get linked channels
    useEffect(() => {
        const linkedChannelsParams: SearchParams = {page: paginationQueryParams.page, per_page: paginationQueryParams.per_page};
        if (searchLinkedChannelsText) {
            linkedChannelsParams.search = searchLinkedChannelsText;
        }

        setGetLinkedChannelsParams(linkedChannelsParams);
        makeApiRequestWithCompletionStatus(pluginApiServiceConfigs.getLinkedChannels.apiServiceName, linkedChannelsParams);
    }, [paginationQueryParams]);

    // Update connected reducer and show alert on successful connection of the user
    useEffect(() => {
        if (connected && !isAlreadyConnected) {
            showAlert({message: 'Your account is connected successfully.', severity: 'default'});
            dispatch(setConnected({connected, msteamsUserId, username, isAlreadyConnected: true}));
        }
    }, [connected, isAlreadyConnected]);

    // Update total linked channels after completion of the api to get linked channels
    useApiRequestCompletionState({
        serviceName: pluginApiServiceConfigs.getLinkedChannels.apiServiceName,
        payload: getLinkedChannelsParams as SearchParams,
        handleSuccess: () => {
            if (linkedChannels) {
                setTotalLinkedChannels([...totalLinkedChannels, ...(linkedChannels as ChannelLinkData[])]);
            }
            if (firstRender && !paginationQueryParams.page) {
                setFirstRender(false);
            }
        },
    });

    // Disconnect user and show alerts on completion of the api to disconnect user
    useApiRequestCompletionState({
        serviceName: pluginApiServiceConfigs.disconnectUser.apiServiceName,
        handleSuccess: () => {
            dispatch(setConnected({connected: false, username: '', msteamsUserId: '', isAlreadyConnected: false}));
            hideDialog();
            showAlert({
                message: 'Your account has been disconnected.',
                severity: 'default',
            });
        },
        handleError: () => {
            showAlert({
                message: 'Error occurred while disconnecting the user.',
                severity: 'error',
            });
            hideDialog();
        },
    });

    useEffect(() => {
        if (refetch) {
            resetStates();
            dispatch(resetRefetch());
        }
    }, [refetch]);

    // Get different states of rhs
    const getRhsView = useCallback(() => {
        // Show spinner in the rhs during loading
        if (isRhsLoading || (firstRender && isLoading)) {
            return (
                <div className='absolute d-flex align-items-center justify-center w-full h-full'>
                    <Spinner size='xl'/>
                </div>
            );
        }

        // Rhs state when user is disconnected and no linked channels are present
        if (!connected && !totalLinkedChannels.length && !searchLinkedChannelsText && !isLoading) {
            return (
                <div className='p-24 d-flex flex-column overflow-y-auto'>
                    <div className='flex-1 d-flex flex-column gap-16 align-items-center my-16'>
                        <div className='d-flex flex-column gap-16 align-items-center'>
                            <Icon
                                width={218}
                                iconName='connectAccount'
                            />
                            <h2 className='text-center wt-600 my-0'>{Constants.connectAccountMsg}</h2>
                        </div>
                        <Button onClick={connectAccount}>{Constants.connectButtonText}</Button>
                    </div>
                    <hr className='w-full my-32'/>
                    <div className='d-flex flex-column gap-24'>
                        <h5 className='my-0 wt-600'>{Constants.listTitle}</h5>
                        <ul className='my-0 px-0 d-flex flex-column gap-20'>
                            {Constants.connectAccountFeatures.map(({icon, text}) => (
                                <li
                                    className='d-flex gap-16 align-items-start'
                                    key={icon}
                                >
                                    <Icon iconName={icon as IconName}/>
                                    <h5 className='my-0 lh-20'>{text}</h5>
                                </li>
                            ))}
                        </ul>
                    </div>
                </div>
            );
        }

        /**
         * Rhs state for the following views:
         * user is disconnected and linked channels are present
         * user is connected and no linked channels are present
         * user is connected and linked channels are present
        */
        return (
            <div className='msteams-sync-rhs flex-1 d-flex flex-column'>
                {connected ? (
                    <div className='py-12 px-20 border-y-1 d-flex gap-8'>
                        <Avatar url={utils.getAvatarUrl(msteamsUserId)}/>
                        <div>
                            <h5 className='my-0 font-12 lh-16'>{'Connected as '}<span className='wt-600'>{username}</span></h5>
                            <Button
                                size='sm'
                                variant='text'
                                className='p-0 lh-16'
                                onClick={() => showDialog({
                                    destructive: true,
                                    primaryButtonText: 'Disconnect',
                                    secondaryButtonText: 'Cancel',
                                    description: 'Are you sure you want to disconnect your Microsoft Teams Account? You will no longer be able to send and receive messages to Microsoft Teams users from Mattermost.',
                                    isLoading: false,
                                    title: 'Disconnect Microsoft Teams Account',
                                })}
                            >{'Disconnect'}</Button>
                        </div>
                        <DialogComponent
                            onSubmitHandler={disconnectUser}
                            onCloseHandler={hideDialog}
                        />
                    </div>
                ) : (
                    <div className='p-20 d-flex flex-column gap-20'>
                        <WarningCard
                            onConnect={connectAccount}
                        />
                    </div>
                )}
                {/* Show spinner during the first load of the linked channels. */}
                {isLoading && firstRender && (
                    <Spinner
                        size='xl'
                        className='scroll-container__spinner mt-10'
                    />
                )}
                {/* State when user is connected, but no linked channels are present. */}
                {!totalLinkedChannels.length && !isLoading && !searchLinkedChannelsText && !previousState?.searchLinkedChannelsText && (
                    <div className='d-flex align-items-center justify-center flex-1 flex-column px-40'>
                        <Icon iconName='noChannels'/>
                        <h3 className='my-0 lh-28 wt-600 text-center'>{'There are no linked channels yet'}</h3>
                        <Button
                            className='mt-16'
                            onClick={() => dispatch(showLinkModal())}
                        >{'Link a Channel'}</Button>
                    </div>
                )}
                {/* State when user is conected and linked channels are present. */}
                {((Boolean(totalLinkedChannels.length) || isLoading || searchLinkedChannelsText || previousState?.searchLinkedChannelsText) && !firstRender) && (
                    <>
                        <div className='d-flex justify-between align-items-center p-20'>
                            <h4 className='font-16 lh-24 my-0 wt-600'>{channelListTitle}</h4>
                            {/* TODO: Replace with Add icon after ui lib version bump */}
                            {connected && (
                                <Button
                                    iconName='Unlink'
                                    size='sm'
                                    onClick={() => dispatch(showLinkModal())}
                                >{'Add'}</Button>
                            )}
                        </div>
                        <div className='p-20 pt-0 my-0'>
                            <Input
                                iconName='MagnifyingGlass'
                                label='Search for a channel'
                                fullWidth={true}
                                value={searchLinkedChannelsText}
                                onChange={(e: React.ChangeEvent<HTMLInputElement>) => setSearchLinkedChannelsText(e.target.value)}
                                onClose={() => setSearchLinkedChannelsText('')}
                            />
                        </div>
                        {/* Show a spinner while searching for a specific linked channel. */}
                        {isLoading && !paginationQueryParams.page ? (
                            <Spinner
                                size='xl'
                                className='scroll-container__spinner'
                            />
                        ) : (
                            <div
                                id='scrollableArea'
                                className='scroll-container flex-1-0-0'
                            >
                                <InfiniteScroll
                                    dataLength={totalLinkedChannels.length}
                                    next={handlePagination}
                                    hasMore={hasMoreLinkedChannels}
                                    loader={<Spinner className='scroll-container__spinner'/>}
                                    endMessage={
                                        <p className='text-center'>
                                            <b>{noMoreChannelsText}</b>
                                        </p>
                                    }
                                    scrollableTarget='scrollableArea'
                                >
                                    {totalLinkedChannels.map(({msTeamsChannelID, ...rest}) => (
                                        <LinkedChannelCard
                                            channelId={msTeamsChannelID}
                                            key={msTeamsChannelID}
                                            {...rest}
                                        />
                                    ))
                                    }
                                </InfiniteScroll>
                            </div>
                        )}
                    </>
                )}
            </div>
        );
    }, [connected, isRhsLoading, isLoading, totalLinkedChannels, firstRender, searchLinkedChannelsText]);

    return (
        <>
            {
                presentInWhitelist ?
                    getRhsView() : 'MS Teams Sync plugin'
            }
            {isOpen && <Snackbar/>}
            {<LinkChannelModal/>}
        </>
    );
};
