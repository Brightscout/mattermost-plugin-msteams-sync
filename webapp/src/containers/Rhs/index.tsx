import React, {useCallback, useEffect, useMemo, useState} from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';

import {Spinner, Tooltip} from '@brightscout/mattermost-ui-library';

import {General as MMConstants} from 'mattermost-redux/constants';

import useApiRequestCompletionState from '../../hooks/useApiRequestCompletionState';
import usePluginApi from '../../hooks/usePluginApi';

import Constants from '../../constants';

import './rhs.scss';

const Rhs = (): JSX.Element => {
    const {state, makeApiRequestWithCompletionStatus, getApiState} = usePluginApi();
    const connected = state.connectedReducer.connected;

    const [totalLinkedChannels, setTotalLinkedChannels] = useState<ChannelLinkData[]>([]);
    const [paginationQueryParams, setPaginationQueryParams] = useState<PaginationQueryParams>({
        page: Constants.DefaultPage,
        per_page: Constants.DefaultPageSize,
    });
    const [getLinkedChannelsParams, setGetLinkedChannelsParams] = useState<PaginationQueryParams | null>(null);

    const connectAccount = useCallback(() => {
        makeApiRequestWithCompletionStatus(Constants.pluginApiServiceConfigs.connect.apiServiceName);
    }, []);

    useEffect(() => {
        const linkedChannelsParams: PaginationQueryParams = {page: paginationQueryParams.page, per_page: paginationQueryParams.per_page};
        setGetLinkedChannelsParams(linkedChannelsParams);
        makeApiRequestWithCompletionStatus(Constants.pluginApiServiceConfigs.getLinkedChannels.apiServiceName, linkedChannelsParams);
    }, [paginationQueryParams]);

    const {data: connectData} = getApiState(Constants.pluginApiServiceConfigs.connect.apiServiceName);
    const {data: linkedChannels, isLoading} = getApiState(Constants.pluginApiServiceConfigs.getLinkedChannels.apiServiceName, getLinkedChannelsParams as PaginationQueryParams);

    useApiRequestCompletionState({
        serviceName: Constants.pluginApiServiceConfigs.connect.apiServiceName,
        handleSuccess: () => {
            if (connectData) {
                window.open((connectData as ConnectData).connectUrl, '_blank');
            }
        },
    });

    useApiRequestCompletionState({
        serviceName: Constants.pluginApiServiceConfigs.getLinkedChannels.apiServiceName,
        payload: getLinkedChannelsParams as PaginationQueryParams,
        handleSuccess: () => {
            if (linkedChannels) {
                setTotalLinkedChannels([...totalLinkedChannels, ...(linkedChannels as ChannelLinkData[])]);
            }
        },
    });

    // Increase the page number by 1
    const handlePagination = () => {
        setPaginationQueryParams({...paginationQueryParams, page: paginationQueryParams.page + 1,
        });
    };

    const hasMoreLinkedChannels = useMemo<boolean>(() => (
        (totalLinkedChannels.length - (paginationQueryParams.page * Constants.DefaultPageSize) === Constants.DefaultPageSize)
    ), [totalLinkedChannels]);

    return (
        <>
            {connected ? (

            // TODO: add disconnect feature later.
                <>{'Connected successfully'}</>
            ) : (
                <div className='msteams-sync-rhs-connect'>
                    <div className='msteams-sync-rhs-connect__heading'>
                        <img
                            className='msteams-sync-rhs-connect__icon'
                            src={Constants.notConnectIconUrl}
                        />
                        <div className='msteams-sync-rhs-connect__body'>
                            <div className='msteams-sync-rhs-connect__title'>{'Please Connect your MS Teams account.'}</div>
                            {'You are not connected to your MS Teams account yet, please connect to your account to continue using Teams sync.'}
                            <div>
                                <button
                                    className='btn btn-primary msteams-sync-rhs-connect__connect-button'
                                    onClick={connectAccount}
                                >
                                    {'Connect Account'}
                                </button>
                            </div>
                        </div>
                    </div>
                </div>

            )}
            <div className='msteams-sync-rhs-divider'/>
            <div className='msteams-sync-rhs-body'>
                <div className='msteams-sync-rhs-body__title'>{'Linked Channels'}</div>
                <div className='msteams-sync-rhs-body__subtitle'>{'Messages will be synchronized between linked channels.'}</div>
                {/* TODO: add search bar later. */}
                {isLoading && !paginationQueryParams.page && <Spinner className='msteams-sync-rhs-body__spinner'/>}
                {totalLinkedChannels.length > 0 && (
                    <div className='link-data__container'>
                        <div className='link-data__title'>
                            <div className='link-data__title-values'>{'Mattermost'}</div>
                            <div className='link-data__title-values'>{'MS Team'}</div>
                        </div>
                        <div
                            id='scrollableArea'
                            className='link-data__container-values'
                        >
                        <InfiniteScroll
                            dataLength={totalLinkedChannels.length}
                            next={handlePagination}
                            hasMore={hasMoreLinkedChannels}
                            loader={<Spinner className='link-data__spinner'/>}
                            endMessage={
                                <p className='text-center'>
                                    <b>{'No more linked channels present.'}</b>
                                </p>
                            }
                            scrollableTarget='scrollableArea'
                        >
                            {totalLinkedChannels.map((link) => (
                                <div
                                    className='link-data'
                                    key={link.msTeamsTeamName}
                                >
                                    <div className='link-data__mm-values'>
                                        <img src={link.mattermostChannelType === MMConstants.PRIVATE_CHANNEL ? Constants.mmPrivateChannelIconUrl : Constants.mmPublicChannelIconUrl}/>
                                        <div className='link-data__body'>
                                            <Tooltip text={link.mattermostChannelName}>
                                                <div className='link-data__channel-name'>
                                                    {link.mattermostChannelName}
                                                </div>
                                            </Tooltip>
                                            <Tooltip text={link.mattermostTeamName}>
                                                <div className='link-data__team-name'>{link.mattermostTeamName}</div>
                                            </Tooltip>
                                        </div>
                                    </div>
                                    <div className='channel-link-icon'>
                                        <img src={Constants.linkIconUrl}/>
                                    </div>
                                    <div className='link-data__ms-values'>
                                        <img src={Constants.msteamsIconUrl}/>
                                        <div className='link-data__body'>
                                            <Tooltip text={link.msTeamsChannelName}>
                                                <div className='link-data__channel-name'>{link.msTeamsChannelName}</div>
                                            </Tooltip>
                                            <Tooltip text={link.msTeamsTeamName}>
                                                <div className='link-data__team-name'>{link.msTeamsTeamName}</div>
                                            </Tooltip>
                                        </div>
                                    </div>
                                    <Tooltip text={'Unlink'}>
                                        <div className='channel-unlink-icon'>
                                            <img
                                                className='channel-unlink-icon__img'

                                                // TODO: Update later
                                                // eslint-disable-next-line no-alert
                                                onClick={() => alert('Unlink chanel')}
                                                src={Constants.channelUnlinkIconUrl}
                                            />
                                        </div>
                                    </Tooltip>
                                </div>
                            ))}
                        </InfiniteScroll>
                        </div>
                    </div>
                )}
                {totalLinkedChannels.length === 0 && !isLoading && (
                    <div className='no-link'>
                        <img src={Constants.globeIconUrl}/>
                        <div className='no-link__title'>{'There are no linked channels'}</div>
                        {connected && (
                            <button
                                className='btn btn-primary'

                                // TODO: Update later
                                // eslint-disable-next-line no-alert
                                onClick={() => alert('open modal!!!!!!!!!')}
                            >
                                {'Link New Channel'}
                            </button>
                        )}
                    </div>
                )}
            </div>
            {connected && totalLinkedChannels.length > 0 && (
                <div className='msteams-sync-rhs-footer'>
                    <div className='msteams-sync-rhs-divider'/>
                    <div className='msteams-sync-rhs-footer__link-btn'>
                        <button
                            className='btn btn-primary'

                            // TODO: Update later
                            // eslint-disable-next-line no-alert
                            onClick={() => alert('open modal!!!!!!!!!')}
                        >
                            {'Link Channel'}
                        </button>
                    </div>
                </div>
            )}
        </>
    );
};

export default Rhs;
