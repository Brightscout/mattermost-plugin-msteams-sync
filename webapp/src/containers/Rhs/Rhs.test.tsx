import React from 'react';

import {RenderResult, render} from '@testing-library/react';

import {Rhs} from './Rhs.container';

let tree: RenderResult;

describe('RHS view', () => {
    beforeEach(() => {
        tree = render(<Rhs/>);
    });

    it('should render correctly', () => {
        expect(tree).toMatchSnapshot();
    });

    it('should render disconnect account button', () => {
        const disconnectButton = tree.getByText('Disconnect');
        expect(disconnectButton).toBeVisible();
    });
});
