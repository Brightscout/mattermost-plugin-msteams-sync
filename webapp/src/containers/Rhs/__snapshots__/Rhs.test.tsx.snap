// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RHS view Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="msteams-sync-utils"
      >
        <div
          class="flex-1 msteams-sync-rhs d-flex flex-column"
        >
          <div
            class="py-12 px-20 border-y-1 d-flex gap-8"
          >
            <div
              style="height: 32px; width: 32px; border-radius: 50%;"
            >
              <img
                src="/plugins/com.mattermost.msteams-sync/avatar/mock_msteams_user_id"
                style="border-radius: 50%;"
              />
            </div>
            <div>
              <h5
                class="my-0 font-12 lh-16"
              >
                Connected as 
                <span
                  class="wt-600"
                >
                  John Doe
                </span>
              </h5>
              <button
                class="sc-kEqXSa fSZMRH mm-button p-0 lh-16"
                width="fit-content"
              >
                Disconnect
              </button>
            </div>
          </div>
          <div
            class="d-flex align-items-center justify-center flex-1 flex-column px-40"
          />
        </div>
      </div>
      <div
        class="msteams-sync-utils"
      >
        <div
          class="fixed bottom-20 right-20 left-20 py-8 px-12 rounded-4 d-flex gap-8 align-items-center justify-between elevation-2 msteams-sync-rhs__snackbar bg-error"
        >
          <div
            class="d-flex align-items-center gap-8"
          >
            <span
              class="icon-white icon-16"
              style="line-height: 0;"
            >
              <svg
                fill="none"
                height="20"
                viewBox="0 0 22 20"
                width="22"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20.648 14.5839C20.904 15.0479 21.024 15.5439 21.008 16.0719C20.992 16.5999 20.848 17.0959 20.576 17.5599C20.304 18.0079 19.936 18.3599 19.472 18.6159C19.024 18.8719 18.536 18.9999 18.008 18.9999H4.016C3.472 18.9999 2.968 18.8719 2.504 18.6159C2.056 18.3599 1.696 18.0079 1.424 17.5599C1.152 17.0959 1.008 16.5999 0.992 16.0719C0.976 15.5439 1.096 15.0479 1.352 14.5839L8.36 1.5759C8.616 1.0799 8.976 0.695901 9.44 0.423902C9.92 0.135902 10.44 -0.00809813 11 -0.00809813C11.56 -0.00809813 12.072 0.135902 12.536 0.423902C13.016 0.695901 13.384 1.0799 13.64 1.5759L20.648 14.5839ZM18.008 17.0079C18.248 17.0079 18.456 16.9359 18.632 16.7919C18.808 16.6319 18.92 16.4399 18.968 16.2159C19.016 15.9759 18.984 15.7439 18.872 15.5199L11.888 2.5359C11.792 2.3599 11.664 2.2319 11.504 2.1519C11.344 2.0559 11.176 2.0079 11 2.0079C10.824 2.0079 10.656 2.0559 10.496 2.1519C10.336 2.2319 10.208 2.3599 10.112 2.5359L3.128 15.5199C3.016 15.7439 2.984 15.9759 3.032 16.2159C3.08 16.4399 3.192 16.6319 3.368 16.7919C3.544 16.9359 3.752 17.0079 3.992 17.0079H18.008ZM11.504 11.0079L12.008 5.0079H9.992L10.496 11.0079H11.504ZM12.008 14.0079C12.008 13.7359 11.904 13.5039 11.696 13.3119C11.504 13.1039 11.272 12.9999 11 12.9999C10.728 12.9999 10.488 13.1039 10.28 13.3119C10.088 13.5039 9.992 13.7359 9.992 14.0079C9.992 14.2799 10.088 14.5119 10.28 14.7039C10.488 14.8959 10.728 14.9919 11 14.9919C11.272 14.9919 11.504 14.8959 11.696 14.7039C11.904 14.5119 12.008 14.2799 12.008 14.0079Z"
                  fill="#D24B4E"
                />
              </svg>
            </span>
            <h5
              class="my-0 lh-24 wt-600 text-white"
            >
              mockMessage
            </h5>
          </div>
          <button
            class="sc-kEqXSa bjmVZZ mm-button snackbar__close"
            width="fit-content"
          >
            <span
              class="icon-white icon-16"
              style="line-height: 0;"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.256 3.80605L9.062 8.00005L13.256 12.194L12.194 13.256L8 9.06205L3.806 13.256L2.744 12.194L6.938 8.00005L2.744 3.80605L3.806 2.74405L8 6.93805L12.194 2.74405L13.256 3.80605Z"
                  fill="#3F4350"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="msteams-sync-utils"
    >
      <div
        class="flex-1 msteams-sync-rhs d-flex flex-column"
      >
        <div
          class="py-12 px-20 border-y-1 d-flex gap-8"
        >
          <div
            style="height: 32px; width: 32px; border-radius: 50%;"
          >
            <img
              src="/plugins/com.mattermost.msteams-sync/avatar/mock_msteams_user_id"
              style="border-radius: 50%;"
            />
          </div>
          <div>
            <h5
              class="my-0 font-12 lh-16"
            >
              Connected as 
              <span
                class="wt-600"
              >
                John Doe
              </span>
            </h5>
            <button
              class="sc-kEqXSa fSZMRH mm-button p-0 lh-16"
              width="fit-content"
            >
              Disconnect
            </button>
          </div>
        </div>
        <div
          class="d-flex align-items-center justify-center flex-1 flex-column px-40"
        />
      </div>
    </div>
    <div
      class="msteams-sync-utils"
    >
      <div
        class="fixed bottom-20 right-20 left-20 py-8 px-12 rounded-4 d-flex gap-8 align-items-center justify-between elevation-2 msteams-sync-rhs__snackbar bg-error"
      >
        <div
          class="d-flex align-items-center gap-8"
        >
          <span
            class="icon-white icon-16"
            style="line-height: 0;"
          >
            <svg
              fill="none"
              height="20"
              viewBox="0 0 22 20"
              width="22"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20.648 14.5839C20.904 15.0479 21.024 15.5439 21.008 16.0719C20.992 16.5999 20.848 17.0959 20.576 17.5599C20.304 18.0079 19.936 18.3599 19.472 18.6159C19.024 18.8719 18.536 18.9999 18.008 18.9999H4.016C3.472 18.9999 2.968 18.8719 2.504 18.6159C2.056 18.3599 1.696 18.0079 1.424 17.5599C1.152 17.0959 1.008 16.5999 0.992 16.0719C0.976 15.5439 1.096 15.0479 1.352 14.5839L8.36 1.5759C8.616 1.0799 8.976 0.695901 9.44 0.423902C9.92 0.135902 10.44 -0.00809813 11 -0.00809813C11.56 -0.00809813 12.072 0.135902 12.536 0.423902C13.016 0.695901 13.384 1.0799 13.64 1.5759L20.648 14.5839ZM18.008 17.0079C18.248 17.0079 18.456 16.9359 18.632 16.7919C18.808 16.6319 18.92 16.4399 18.968 16.2159C19.016 15.9759 18.984 15.7439 18.872 15.5199L11.888 2.5359C11.792 2.3599 11.664 2.2319 11.504 2.1519C11.344 2.0559 11.176 2.0079 11 2.0079C10.824 2.0079 10.656 2.0559 10.496 2.1519C10.336 2.2319 10.208 2.3599 10.112 2.5359L3.128 15.5199C3.016 15.7439 2.984 15.9759 3.032 16.2159C3.08 16.4399 3.192 16.6319 3.368 16.7919C3.544 16.9359 3.752 17.0079 3.992 17.0079H18.008ZM11.504 11.0079L12.008 5.0079H9.992L10.496 11.0079H11.504ZM12.008 14.0079C12.008 13.7359 11.904 13.5039 11.696 13.3119C11.504 13.1039 11.272 12.9999 11 12.9999C10.728 12.9999 10.488 13.1039 10.28 13.3119C10.088 13.5039 9.992 13.7359 9.992 14.0079C9.992 14.2799 10.088 14.5119 10.28 14.7039C10.488 14.8959 10.728 14.9919 11 14.9919C11.272 14.9919 11.504 14.8959 11.696 14.7039C11.904 14.5119 12.008 14.2799 12.008 14.0079Z"
                fill="#D24B4E"
              />
            </svg>
          </span>
          <h5
            class="my-0 lh-24 wt-600 text-white"
          >
            mockMessage
          </h5>
        </div>
        <button
          class="sc-kEqXSa bjmVZZ mm-button snackbar__close"
          width="fit-content"
        >
          <span
            class="icon-white icon-16"
            style="line-height: 0;"
          >
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.256 3.80605L9.062 8.00005L13.256 12.194L12.194 13.256L8 9.06205L3.806 13.256L2.744 12.194L6.938 8.00005L2.744 3.80605L3.806 2.74405L8 6.93805L12.194 2.74405L13.256 3.80605Z"
                fill="#3F4350"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
